### DISCUSSION TITLE
`GM` APIs Usage Recommendation

### DISCUSSION THREAD

**Discussion by Malix-Labs** [Tampermonkey/tampermonkey#1745] [Q&A] [answered]
Hello!

How would you recommend using `GM` APIs: is it better to use `GM_*` or `GM.*`?
What are the benefit of one on the other?

Thanks!

**Comment by 7nik** [comment, marked as answer]
When only one choice is possible:
a) you want your userscript to be compatible with Greasemonkey - it supports only `GM.*` (but it supports only a few of them);
b) the userscript must run at `document-start`, read settings or some other data and do something synchronously - `GM_*`.

The only difference for the rest cases is that for `GM.*`, you need to add `await` and often declare functions as `async`. Note, now the top-level await is supported.

Technically, `GM.*` may have an implementation with less impact on the page-loading performance. Still, on the other hand, it'd be weird to have two separate implementations of sync and async API doing the same thing. Anyway, I don't know much about this and don't want to speculate.

**Reply by Malix-Labs** [reply]
> the userscript must run at `document-start`

Why can't you do that with `GM.*`?

> read settings

What do you mean by that? Which settings and why can't you do that with `GM.*`?

> do something synchronously

`GM.*` can do that too with await

**Reply by 7nik** [reply]
You cannot do `let someOption = await GM.getValue("some option name");` synchronously.

**Reply by Malix-Labs** [reply]
> You cannot do `let someOption = await GM.getValue("some option name");` synchronously.

what do you mean by synchronously then?

**Reply by 7nik** [reply]
`let someOption = GM_getValue("some option name");` - no `await`, no delay, no way for other (page's) code to be executed - there are cases when userscript must complete initialization before page scripts run.

**Reply by Malix-Labs** [reply]
I wonder, can you use `close` with `GM_openInTab` for example?
I would like to use information stored on another page to use in the main page.
the ideal would be that the opened tab get run in the background, without focus

**Reply by 7nik** [reply]
If you want to scrape some info from a page, using GM_XHR to get the HTML and extract the info is better. If the info is loaded dynamically, use GM_XHR to access the info's endpoint.

**Reply by Malix-Labs** [reply]
> If you want to scrape some info from a page, using GM_XHR to get the HTML and extract the info is better. If the info is loaded dynamically, use GM_XHR to access the info's endpoint.

Thanks for that suggestion!

***

Still, do you know how to do

> use `close` with `GM_openInTab` for example

**Reply by 7nik** [reply]
According to the docs, it should be like this:
```js
const tab = GM_openInTab("https://www.example.com/");

// latter
tab.close();

// somewhere else
if (tab.closed) {
  // tab is closed
  // maybe you want to re-open in
  // or complain to the user about the closed tab
} else {
  doSomeAction();
}

// and this can be done too
tab.addEventListener("close", (event) => {
  //I dunno if `event` will contain something useful. You can log it to check out
  doSomeActionOnClose(); - // there various popups that get self-closed
});
```
